tests/e2e/common/diagnosticcollector.go

```diff
@@ -0,0 +1,465 @@
+	if err != nil {
+		return err
+	}
+	defer nfrsFile.Close()
+
+	if err = c.writeToFile(nfrsFile, nfrs); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createKubernetesClient(kubeconfig string) (*kubernetes.Clientset, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientset, nil
+}
+
+type DiagnosticCollector struct {
+	*Config
+	collectors []Collector
+}
+
+func WithNamespace(namespace string) func(*DiagnosticCollector) {
+	return func(d *DiagnosticCollector) {
+		d.Config.namespace = namespace
+	}
+}
+
+func WithArtifactDir(artifactDir string) func(*DiagnosticCollector) {
+	return func(d *DiagnosticCollector) {
+		d.Config.artifactDir = artifactDir
+	}
+}
+
+func WichNFDClient(nfdClient *nfdclient.Clientset) func(*DiagnosticCollector) {
+	return func(d *DiagnosticCollector) {
+		d.NfdClient = nfdClient
+	}
+}
+
+func WithObjects(objects ...string) func(*DiagnosticCollector) {
+	// First let's remove duplicates
+	objectsMap := make(map[string]struct{})
+	for _, obj := range objects {
+		objectsMap[obj] = struct{}{}
+	}
+
+	return func(d *DiagnosticCollector) {
+		for obj := range objectsMap {
+			switch obj {
+			case Nodes:
+				d.Add(nodes{Config: d.Config})
+			case Namespaces:
+				d.Add(namespaces{Config: d.Config})
+			case Pods:
+				d.Add(pods{Config: d.Config})
+			case Deployments:
+				d.Add(deployments{Config: d.Config})
+			case DaemonSets:
+				d.Add(daemonsets{Config: d.Config})
+			case Jobs:
+				d.Add(jobs{Config: d.Config})
+			case NodeFeature:
+				d.Add(nodeFeatures{Config: d.Config, NfdClient: d.NfdClient})
+			case NodeFeatureRule:
+				d.Add(nodeFeatureRules{Config: d.Config, NfdClient: d.NfdClient})
+			}
+		}
+	}
+}
+
+// rewrite the NewDiagnosticCollector function to accept a variadic function
+func NewDiagnosticCollector(opts ...func(*DiagnosticCollector)) (*DiagnosticCollector, error) {
+	c := &Config{}
+	dc := &DiagnosticCollector{
+		Config: c,
+	}
+
+	// use the variadic function to set the options
+	for _, opt := range opts {
+		opt(dc)
+	}
+
+	// Create the artifact directory
+	if err := os.MkdirAll(filepath.Join(c.artifactDir, c.namespace), os.ModePerm); err != nil {
+		return dc, err
+	}
+
+	// Get the kubeconfig
+	kubeconfig := os.Getenv("KUBECONFIG")
+
+	// Create the Kubernetes client
+	clientset, err := createKubernetesClient(kubeconfig)
+	if err != nil {
+		return dc, fmt.Errorf("error creating kubernetes client: %v", err)
+	}
+	c.clientset = clientset
+
+	return dc, nil
+}
+
+func (d *DiagnosticCollector) Add(f Collector) {
+	d.collectors = append(d.collectors, f)
+}
+
+func (d *DiagnosticCollector) Collect(ctx context.Context) error {
+	// Redirect stdout and stderr to logs
+	logFile, err := d.createFile("diagnostic_collector.stdout.log")
+	if err != nil {
+		return err
+	}
+	defer logFile.Close()
+	d.log = logFile
```

Comment:
>Question: Do we need a separate log and error log?
we are printing 2 files, one for stdout and one for stderr. 

For the rest of your comment: We don't wanna do that, since this tool is mostly to generate files (artifacts) that can later be used to debug a failed test run